.github/workflows/staging-deployment.yml

```diff
@@ -0,0 +1,179 @@
+        with:
+          name: Artemis.war
+          path: artifacts
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          run-id: ${{ env.WORKFLOW_RUN_ID  }}
+
+      - name: Setup SSH
+        env:
+          DEPLOYMENT_SSH_KEY: ${{ secrets.DEPLOYMENT_SSH_KEY }}
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
+          DEPLOYMENT_HOST_PUBLIC_KEYS: ${{ vars.DEPLOYMENT_HOST_PUBLIC_KEYS }}
+        run: |
+          mkdir -p ~/.ssh
+          ssh-agent -a $SSH_AUTH_SOCK > /dev/null
+          ssh-add - <<< "$DEPLOYMENT_SSH_KEY"
+          cat - <<< $DEPLOYMENT_HOST_PUBLIC_KEYS >> ~/.ssh/known_hosts
+          ssh-add -l || echo "No identities found"
+
+      - name: Phase 1 - Stop Secondary Nodes
+        run: |
+          for node in $DEPLOYMENT_HOSTS_SECONDARY; do
+            echo "[INFO] Stop artemis.service on ${node} ..."
+            ssh -o BatchMode=yes -o StrictHostKeyChecking=accept-new \
+                -l $DEPLOYMENT_USER $node "sudo systemctl stop artemis" || {
+              echo "::error::Failed to stop service on ${node}"
+              exit 1
+            }   
+          done
+
+      - name: Phase 1 - Deploy to Primary Node
+        run: |
+          echo "[INFO] Deploy on $DEPLOYMENT_HOSTS_PRIMARY ..."
+          
+          # Remove old backup
+          ssh -o BatchMode=yes -o StrictHostKeyChecking=accept-new \
+              -l $DEPLOYMENT_USER $DEPLOYMENT_HOSTS_PRIMARY \
+              "rm -f $DEPLOYMENT_FOLDER/artemis.war.old" || exit 1
+          
+          # Copy new war file
+          scp -o BatchMode=yes -o StrictHostKeyChecking=accept-new \
+              artifacts/*.war $DEPLOYMENT_USER@$DEPLOYMENT_HOSTS_PRIMARY:$DEPLOYMENT_FOLDER/artemis.war.new || {
+            echo "::error::Failed to copy war file to primary host"
+            exit 1
+          }
+
+          # Stop service, rotate files, and start service
+          ssh -o BatchMode=yes -o StrictHostKeyChecking=accept-new \
+              -l $DEPLOYMENT_USER $DEPLOYMENT_HOSTS_PRIMARY << 'ENDSSH' || exit 1
+            set -e
+            sudo systemctl stop artemis
+            mv $DEPLOYMENT_FOLDER/artemis.war $DEPLOYMENT_FOLDER/artemis.war.old
+            mv $DEPLOYMENT_FOLDER/artemis.war.new $DEPLOYMENT_FOLDER/artemis.war
+            sudo systemctl start artemis
+          ENDSSH
+
+      - name: Verify Primary Node Deployment
+        id: verify_deployment
+        run: |
+          while true; do
+            echo "Performing health check..."
+      
+            RESPONSE=$(curl -s -f $HEALTH_CHECK_URL || echo '{"status":"DOWN"}')
+            STATUS=$(echo $RESPONSE | jq -r '.status')
+      
+            if [ "$STATUS" = "UP" ]; then
+              echo "Health check passed! Application is UP"
+              exit 0
+            else
+              echo "Health check failed. Status: $STATUS"
+              echo "Waiting 10 seconds before next attempt..."
+              sleep 10
+            fi
+          done
```

Comment:
Optional: I'd add a timeout specifically for this step. Would be more accurate than a job-level timeout